lib/include/atomic.h

```diff
@@ -0,0 +1,484 @@
+    {
+        *pDestination = ulExchange;
+        bExecutionStatus = true;
+    }
+
+    ATOMIC_EXIT_CRITICAL( temp );
+
+    return bExecutionStatus;
+
+#endif
+
+}
+
+/**
+ * Atomic swap (pointers)
+ *
+ * @brief Atomically sets the address pointed to by *ppDestination to the value of *pExchange.
+ *
+ * @param[in, out] ppDestination  Pointer to memory location from where a pointer value is to be loaded and writen back to.
+ * @param[in] pExchange           Pointer value to be written to *ppDestination.
+ *
+ * @return The initial value of *ppDestination.
+ */
+static portFORCE_INLINE void * Atomic_SwapPointers_p32( void * volatile * ppDestination, void * pExchange )
+{
+    void * pReturnValue;
+
+#if defined ( configUSE_ATOMIC_INSTRUCTION ) && ( configUSE_ATOMIC_INSTRUCTION == 1 )
+
+    __atomic_exchange( ppDestination, &pExchange, &pReturnValue, __ATOMIC_SEQ_CST );
+
+#else
+
+    CriticalSessionType_t temp;
+
+    temp = ATOMIC_ENTER_CRITICAL( );
+
+    pReturnValue = *ppDestination;
+
+    *ppDestination = pExchange;
+
+    ATOMIC_EXIT_CRITICAL( temp );
+
+#endif
+
+    return pReturnValue;
+}
+
+/**
+ * Atomic compare-and-swap (pointers)
+ *
+ * @brief Performs an atomic compare-and-swap operation on the specified pointer values.
+ *
+ * @param[in, out] ppDestination Pointer to memory location from where a pointer value is to be loaded and checked.
+ * @param[in] pExchange          If condition meets, write this value to memory.
+ * @param[in] pComparand         Swap condition, checks and waits for *ppDestination to be equal to *pComparand.
+ *
+ * @return bool True for swapped, false for not swapped.
+ *
+ * @note This function only swaps *ppDestination with pExchange, if previous *ppDestination value equals pComparand.
+ */
+static portFORCE_INLINE bool Atomic_CompareAndSwapPointers_p32( void * volatile * ppDestination, void * pExchange, void * pComparand )
+{
+
+#if defined ( configUSE_ATOMIC_INSTRUCTION ) && ( configUSE_ATOMIC_INSTRUCTION == 1 )
+    return __atomic_compare_exchange( ppDestination, &pComparand, &pExchange, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST );
+
+#else
+
+    bool bExecutionStatus = false;
+    CriticalSessionType_t temp;
+
+    temp = ATOMIC_ENTER_CRITICAL( );
+
+    if ( *ppDestination == pComparand )
+    {
+        *ppDestination = pExchange;
+        bExecutionStatus = true;
+    }
+
+    ATOMIC_EXIT_CRITICAL( temp );
+    return bExecutionStatus;
+
+#endif
+
+}
+
+
+/*----------------------------- Arithmetic ------------------------------*/
+
+/**
+ * Atomic add
+ *
+ * @brief Adds count to the value of the specified 32-bit variable as an atomic operation.
+ *
+ * @param[in,out] pAddend  Pointer to memory location from where value is to be loaded and written back to.
+ * @param[in] lCount       Value to be added to *pAddend.
+ *
+ * @return previous *pAddend value.
+ */
+static portFORCE_INLINE int32_t Atomic_Add_i32( int32_t volatile * pAddend, int32_t lCount )
```

Comment:
I thought we decided to do all uint32_t (unsigned) first... I believe signed may be just as good, just confirming...

Reply:
Per offline discussion, updating to uint32_t for all APIs. And it's up to user to handle overflow/underflow correctly. 

C99 standard, thanks to Cobus.
```
6.3.1.3 Signed and unsigned integers

1 When a value with integer type is converted to another integer type other than _Bool, if the value can be represented by the new type, it is unchanged.

2 Otherwise, if the new type is unsigned, the value is converted by repeatedly adding or subtracting one more than the maximum value that can be represented in the new type until the value is in the range of the new type.49)

3 Otherwise, the new type is signed and the value cannot be represented in it; either the result is implementation-defined or an implementation-defined signal is raised.
```